### PR TITLE
able to run environment-based fun-untyped with Java engine

### PR DESCRIPTION
**Pull request by yilongli**
Sorry, looks like I have completely messed up the commit history... I tried to sync my fork with the upstream repository and somehow I merged my local repo with my fork. Then a lot up duplicate commits show up. I guess that's because I accidentally changed a commit that has been pushed.

### CODE REVIEW

**Comment by dwightguth** [comment]
Yeah I suggest you save your changes somewhere, create a clean repository that contains only the upstream commit history, and then cherry-pick only the changes you want reviewed onto the fork, and push it again. I can't approve a pull request that contains changes belonging to someone else which have already been pushed.

**Comment by yilongli** [comment]
I have rewritten the history and the `ant test` has passed. It should be ready for review now.

File: src/javasources/KTool/src/org/kframework/backend/java/builtins/BuiltinFloatOperations.java [hunk 1]
```diff
@@ -29,7 +29,7 @@ public static Term sub(Term term1, Term term2, TermContext context) {
             return new KItem(
                     KLabelConstant.of("'--Float_",context1),
                     new KList(ImmutableList.<Term>of(term2),null),
-                    context1).evaluate(context);
```
Comment (dwightguth) [inline src/javasources/KTool/src/org/kframework/backend/java/builtins/BuiltinFloatOperations.java:32]: I'm not sure I'm comfortable with KItem and KLabelConstant taking different Context classes.

File: src/javasources/KTool/src/org/kframework/backend/java/builtins/MetaK.java [hunk 2]
```diff
@@ -36,13 +45,35 @@ public static BoolToken unifiable(Term term1, Term term2, TermContext context) {
         SymbolicUnifier unifier = new SymbolicUnifier(constraint, context);
         try {
             unifier.unify(term1, term2);
-        } catch (MatcherException e) {
+        } catch (UnificationFailure e) {
```
Comment (dwightguth) [inline src/javasources/KTool/src/org/kframework/backend/java/builtins/MetaK.java:47]: why are these two different exception types?

File: src/javasources/KTool/src/org/kframework/backend/java/builtins/MetaK.java [hunk 3]
```diff
@@ -36,13 +45,35 @@ public static BoolToken unifiable(Term term1, Term term2, TermContext context) {
...
+     * @param context
+     *            the term context
+     * @return {@link BoolToken#TRUE} if the two terms can be matched;
+     *         otherwise, {@link BoolToken#FALSE}
+     */
+    public static BoolToken matchable(Term subject, Term pattern, TermContext context) {
+        PatternMatcher matcher = new PatternMatcher(context);
+        try {
+            matcher.match(subject, pattern);
+        } catch (PatternMatchingFailure e) {
```
Comment (dwightguth) [inline src/javasources/KTool/src/org/kframework/backend/java/builtins/MetaK.java]: Same question, why a third Exception class?

File: src/javasources/KTool/src/org/kframework/backend/java/indexing/KLabelIndex.java [hunk 4]
```diff
@@ -14,9 +14,16 @@ public KLabelIndex(KLabelConstant kLabel) {
...
+        return kLabel;
+    }
+    
     @Override
     public boolean isUnifiable(Index index) {
-        return index instanceof TopIndex || equals(index);
+        return index instanceof TopIndex
+                || equals(index)
+                || (index instanceof TokenIndex && 
+                        kLabel.sorts().contains(((TokenIndex) index).sort()));
```
Comment (dwightguth) [inline src/javasources/KTool/src/org/kframework/backend/java/indexing/KLabelIndex.java]: what exactly is this doing?

File: src/javasources/KTool/src/org/kframework/backend/java/indexing/TokenIndex.java [hunk 5]
```diff
@@ -10,10 +11,17 @@
...
+        return sort;
+    }
 
     @Override
     public boolean isUnifiable(Index index) {
-        return index instanceof TopIndex || equals(index);
+        return index instanceof TopIndex
+                || equals(index)
+                || (index instanceof KLabelIndex && 
+                        ((KLabelIndex) index).kLabel().sorts().contains(sort));
```
Comment (dwightguth) [inline src/javasources/KTool/src/org/kframework/backend/java/indexing/TokenIndex.java]: same question

File: src/javasources/KTool/src/org/kframework/backend/java/kil/KItem.java [hunk 6]
```diff
@@ -67,7 +73,21 @@ public KItem(Term kLabel, Term kList, Context context) {
             List<Production> productions = kLabelConstant.productions();
             if (productions.size() != 0) {
                 Set<String> sorts = new HashSet<String>();
-
+                
+                if (!K.do_kompilation) {
+                    /* YilongL: user-defined sort predicate rules are interpreted as overloaded productions at runtime */
```
Comment (dwightguth) [inline src/javasources/KTool/src/org/kframework/backend/java/kil/KItem.java:87]: can you reword this to make it clearer? something like: sort checks are not implemented as rewrite rules in Java, so we need to precompute the sort of terms.

File: src/javasources/KTool/src/org/kframework/backend/java/builtins/BuiltinFloatOperations.java (hunk 1 above)
Comment (yilongli) [inline src/javasources/KTool/src/org/kframework/backend/java/builtins/BuiltinFloatOperations.java:32]: I agree. I tend to change KLabelConstant to receive TermContext as well.

File: src/javasources/KTool/src/org/kframework/backend/java/builtins/MetaK.java (hunk 2 above)
Comment (yilongli) [inline src/javasources/KTool/src/org/kframework/backend/java/builtins/MetaK.java:47]: MatcherException is from the generic kil class hierarchy. It is relatively expensive to construct because it needs to fill in the stacktrace. Since the Java engine only uses this exception for control-flow, I wrote the UnificationFailure/PatternMatchingFailure which implements the singleton pattern and has no stacktrace inside.

File: src/javasources/KTool/src/org/kframework/backend/java/builtins/MetaK.java (hunk 3 above)
Comment (yilongli) [inline src/javasources/KTool/src/org/kframework/backend/java/builtins/MetaK.java]: I can combine UnificationFailure and PatternMatchingFailure but I prefer to keep it this way.

Comment (dwightguth) [inline src/javasources/KTool/src/org/kframework/backend/java/builtins/MetaK.java]: I'm fine with separate exceptions for these two if we make a common base
class so we can catch them together if we ever want to.

File: src/javasources/KTool/src/org/kframework/backend/java/kil/KItem.java [hunk 7]
```diff
@@ -181,8 +201,32 @@ public KItem(Term kLabel, Term kList, Context context) {
...
     }
+    
+    public boolean isEvaluable(TermContext context) {
+        if (evaluable != null) {
+            return evaluable;
+        }
+
+        evaluable = false;
+        if (!(kLabel instanceof KLabelConstant)) {
+            return evaluable;
```
Comment (dwightguth) [inline src/javasources/KTool/src/org/kframework/backend/java/kil/KItem.java]: return false

File: src/javasources/KTool/src/org/kframework/backend/java/kil/KItem.java [hunk 8]
```diff
@@ -181,8 +201,32 @@ public KItem(Term kLabel, Term kList, Context context) {
...
+        }
+        KLabelConstant kLabelConstant = (KLabelConstant) kLabel;
+        
+        if (!(kList instanceof KList)) {
+            return evaluable;
+        }
```
Comment (dwightguth) [inline src/javasources/KTool/src/org/kframework/backend/java/kil/KItem.java:230]: return false

File: src/javasources/KTool/src/org/kframework/backend/java/kil/KItem.java [hunk 9]
```diff
@@ -181,8 +201,32 @@ public KItem(Term kLabel, Term kList, Context context) {
...
+        
+        if (kLabelConstant.label().startsWith("is")
+                || !context.definition().functionRules().get(kLabelConstant).isEmpty()
+                || BuiltinFunction.isBuiltinKLabel(kLabelConstant)) {
+            return evaluable = true;
```
Comment (dwightguth) [inline src/javasources/KTool/src/org/kframework/backend/java/kil/KItem.java]: evaluable = true

File: src/javasources/KTool/src/org/kframework/backend/java/indexing/KLabelIndex.java (hunk 4 above)
Comment (yilongli) [inline src/javasources/KTool/src/org/kframework/backend/java/indexing/KLabelIndex.java]: Consider the following productions and the lookup rule in fun-untyped:
`syntax Name ::= Token{[a-z][\_a-zA-Z0-9]*}       [notInRules]`
`syntax Name ::= "$h"`

```
  rule <k> X:Name => V ...</k>
       <env>... X |-> L ...</env>
       <store>... L |-> V ...</store>
```

Name is a token sort thanks to the first production and, therefore, the lookup rule has a `TokenIndex`. However, `$h` has a `KLabelIndex` and will get stuck when it appears on top of the k cell.

File: src/javasources/KTool/src/org/kframework/backend/java/kil/KLabelConstant.java [hunk 10]
```diff
@@ -43,6 +50,12 @@ private KLabelConstant(String label, Context context) {
         this.label = label;
         productions = ImmutableList.copyOf(context.productionsOf(label));
         
+        Set<String> set = new HashSet<>();
+        for (Production prod : productions) {
+            set.add(prod.getSort());
+        }
+        sorts = ImmutableSet.copyOf(set);
```
Comment (dwightguth) [inline src/javasources/KTool/src/org/kframework/backend/java/kil/KLabelConstant.java]: is there an ImmutableSet builder?

File: src/javasources/KTool/src/org/kframework/backend/java/symbolic/KILtoBackendJavaKILTransformer.java [hunk 11]
```diff
@@ -141,7 +141,7 @@ public ASTNode transform(org.kframework.kil.KApp node) throws TransformerExcepti
         if (kList instanceof Variable) {
             kList = new KList((Variable) kList);
         }
-        return new KItem(kLabel, kList, this.context);
+        return new KItem(kLabel, kList, TermContext.of(definition));
```
Comment (dwightguth) [inline src/javasources/KTool/src/org/kframework/backend/java/symbolic/KILtoBackendJavaKILTransformer.java:172]: is this method guaranteed to return the same object each time?

File: src/javasources/KTool/src/org/kframework/backend/java/symbolic/SubstitutionTransformer.java [hunk 12]
```diff
@@ -32,7 +32,29 @@ public SubstitutionTransformer(Map<Variable, ? extends Term> substitution, TermC
         postTransformer.addTransformer(new LocalSubstitutionTransformer());
         postTransformer.addTransformer(new VariableUpdaterTransformer());
     }
+    
+    /**
+     * Stops further traversing the term when it contains no variable that needs
+     * to be substituted.
+     * 
+     */
+    private class LocalVariableChecker extends LocalTransformer {
```
Comment (dwightguth) [inline src/javasources/KTool/src/org/kframework/backend/java/symbolic/SubstitutionTransformer.java:41]: why was this class moved?

File: src/javasources/KTool/src/org/kframework/compile/transformers/AddInjections.java [hunk 13]
```diff
@@ -192,7 +192,7 @@ public ASTNode transform(TermCons node) throws TransformerException {
         if (sort.equals(KSorts.K) || sort.equals(KSorts.KLABEL) || sort.equals(KSorts.KLIST)) {
             transformedNode.setSort(KSorts.KITEM);
             // TODO (AndreiS): remove special case
-            if (node.getProduction().getLabel().equals("#if_#then_#else_#fi")) {
+            if (node.getProduction().getLabel().equals("#if_#then_#else_#fi") && !sort.equals(KSorts.KLIST)) {
```
Comment (dwightguth) [inline src/javasources/KTool/src/org/kframework/compile/transformers/AddInjections.java:195]: I'm kind of uncomfortable with this being special cased. is there a more uniform way of doing this?

File: tests/regression/java-rewrite-engine/smt_model/program.test.out [hunk 14]
```diff
@@ -1,3 +1,3 @@
 <k>
-    _0!0:Int |-> 1
```
Comment (dwightguth) [inline tests/regression/java-rewrite-engine/smt_model/program.test.out:2]: this doesn't seem connected? why is this a part of this diff?

File: src/javasources/KTool/src/org/kframework/backend/java/indexing/KLabelIndex.java (hunk 4 above)
Comment (dwightguth) [inline src/javasources/KTool/src/org/kframework/backend/java/indexing/KLabelIndex.java]: Okay, but this doesn't feel scalable. Are we going to have an or clause for
every pairwise set of kinds? Can't we handle this more generically somehow?

Also, do we even want to support this use case? It feels wrong to me to
have the ability to overload token sorts with arbitrary productions.

File: src/javasources/KTool/src/org/kframework/backend/java/symbolic/KILtoBackendJavaKILTransformer.java (hunk 11 above)
Comment (yilongli) [inline src/javasources/KTool/src/org/kframework/backend/java/symbolic/KILtoBackendJavaKILTransformer.java:172]: Yes, TermContext objects are cached.

File: src/javasources/KTool/src/org/kframework/backend/java/symbolic/SubstitutionTransformer.java (hunk 12 above)
Comment (yilongli) [inline src/javasources/KTool/src/org/kframework/backend/java/symbolic/SubstitutionTransformer.java:41]: I just reordered the classes to fit the pre/post transformers order.

File: src/javasources/KTool/src/org/kframework/compile/transformers/AddInjections.java (hunk 13 above)
Comment (yilongli) [inline src/javasources/KTool/src/org/kframework/compile/transformers/AddInjections.java:195]: I agree. I am not very sure about the special case here, so I just try to write whatever I need without breaking existing code.
@andreistefanescu Could you shed some light to this case.

File: tests/regression/java-rewrite-engine/smt_model/program.test.out (hunk 14 above)
Comment (yilongli) [inline tests/regression/java-rewrite-engine/smt_model/program.test.out:2]: `_0!0` is a generated fresh variable. The name often gets changed as we change the implementation of the Java engine. In this case, it got changed because I fixed a bug related to detecting unbound variables on the RHS of a rule.

File: src/javasources/KTool/src/org/kframework/backend/java/builtins/MetaK.java (hunk 3 above)
Comment (yilongli) [inline src/javasources/KTool/src/org/kframework/backend/java/builtins/MetaK.java]: Can you come up with a good name for the base class?

File: src/javasources/KTool/src/org/kframework/backend/java/indexing/KLabelIndex.java (hunk 4 above)
Comment (yilongli) [inline src/javasources/KTool/src/org/kframework/backend/java/indexing/KLabelIndex.java]: For now, it is just between TokenIndex and KLabelIndex. I am not sure what would be a better way to handle this but I will keep that in mind.

I am afraid I cannot answer your second question. I would like to get rid of all these tricky cases and keep my code clean of course. @grosu

File: src/javasources/KTool/src/org/kframework/backend/java/builtins/MetaK.java (hunk 3 above)
Comment (dwightguth) [inline src/javasources/KTool/src/org/kframework/backend/java/builtins/MetaK.java]: UnificationOrMatchingFailure

File: tests/regression/java-rewrite-engine/smt_model/program.test.out (hunk 14 above)
Comment (dwightguth) [inline tests/regression/java-rewrite-engine/smt_model/program.test.out:2]: Okay, this is a failure in test design. If the name of the key of this map
is unspecified, then we shouldn't specify it in the test output. is there a
way to specify a k pattern that the output of this program should match?

File: src/javasources/KTool/src/org/kframework/backend/java/indexing/KLabelIndex.java (hunk 4 above)
Comment (dwightguth) [inline src/javasources/KTool/src/org/kframework/backend/java/indexing/KLabelIndex.java]: No it's not though. You've already special cased TopIndex as well. That's
what I'm concerned about here: we seem to be writing an algorithm that
scales quadratically in the number of kinds. And for all I really know
there may be other corner cases you're missing also, since you apparently
only just now thought of this one. so I'm not really very pleased with this
method even if we can remove the corner case in question. If we're doing
things based on sorts but having to do cases kind by kind, what have we
done to solve this in the past? This sounds very similar to something we
should have had to deal with for the unification algorithm itself. How did
we handle it there?

File: src/javasources/KTool/src/org/kframework/compile/transformers/AddInjections.java (hunk 13 above)
Comment (andreistefanescu) [inline src/javasources/KTool/src/org/kframework/compile/transformers/AddInjections.java:195]: The desired signature of if_then_else_fi is forall t . bool -> t -> t -> t. Unfortunately, the K frontend does not support that, so it has the more generic signature bool -> klist -> klist -> klist. This would result in if_then_else_fi being always injected. I don't know how else to avoid this problem.

File: src/javasources/KTool/src/org/kframework/backend/java/indexing/KLabelIndex.java (hunk 4 above)
Comment (andreistefanescu) [inline src/javasources/KTool/src/org/kframework/backend/java/indexing/KLabelIndex.java]: I think
`syntax Name ::= "$h"``

should actually be
`syntax Name ::= Token{$h}``

We should probably have a check in compile for this.

Comment (dwightguth) [inline src/javasources/KTool/src/org/kframework/backend/java/indexing/KLabelIndex.java]: we could actually make this even more general:

syntax A ::= "string"
syntax A ::= Token{regex}

If we wanted to get really crazy, even something like:

syntax A ::= "string"
syntax A ::= `regex`
syntax B ::= "string terminal" NonTerminal `regex terminal`

The semantics would be that if a terminal is all by itself in a production,
the terminal is captured and put in a token, otherwise it's discarded.

Just a thought.

File: src/javasources/KTool/src/org/kframework/backend/java/indexing/TokenIndex.java [hunk 15]
```diff
@@ -10,10 +11,31 @@
...
+                 * therefore, it's possible to match a TokenIndex with a
+                 * KLabelIndex. For instance, consider the following productions
+                 * and rule:
+                 * 
+                 * syntax Name ::= some lexical production
+                 *               | "$h"
+                 *               
+                 * rule X:Name => .      
+                 * 
+                 * The above rule should be able to match when "$h" appears
```
Comment (dwightguth) [inline src/javasources/KTool/src/org/kframework/backend/java/indexing/TokenIndex.java]: this doesn't address my original concern that these methods require a quadratic number of code changes.

File: src/javasources/KTool/src/org/kframework/backend/java/symbolic/KILtoBackendJavaKILTransformer.java [hunk 16]
```diff
@@ -12,13 +12,40 @@
...
 import java.util.Set;
 
-import com.google.common.collect.Iterables;
 import org.kframework.backend.java.builtins.BoolToken;
 import org.kframework.backend.java.builtins.Int32Token;
 import org.kframework.backend.java.builtins.IntToken;
 import org.kframework.backend.java.builtins.StringToken;
 import org.kframework.backend.java.builtins.UninterpretedToken;
-import org.kframework.backend.java.kil.*;
+import org.kframework.backend.java.kil.BuiltinList;
```
Comment (dwightguth) [inline src/javasources/KTool/src/org/kframework/backend/java/symbolic/KILtoBackendJavaKILTransformer.java:20]: I didn't mean remove all star imports. kil is different. this is actually now less readable.

File: src/javasources/KTool/src/org/kframework/backend/java/symbolic/PatternMatcher.java [hunk 17]
```diff
@@ -205,7 +205,7 @@
         return result;
     }
 
-    private PatternMatcher(TermContext context) {
```
Comment (dwightguth) [inline src/javasources/KTool/src/org/kframework/backend/java/symbolic/PatternMatcher.java:208]: why was this done?

File: src/javasources/KTool/src/org/kframework/compile/transformers/AddInjections.java (hunk 13 above)
Comment (dwightguth) [inline src/javasources/KTool/src/org/kframework/compile/transformers/AddInjections.java:195]: what's stopping me from writing my own if then else fi and having it break because this special case doesn't apply?

**Comment by dwightguth** [comment]
I still can't merge this until I get answers to the rest of my questions.

File: src/javasources/KTool/src/org/kframework/compile/transformers/AddInjections.java (hunk 13 above)
Comment (andreistefanescu) [inline src/javasources/KTool/src/org/kframework/compile/transformers/AddInjections.java:195]: Technically, it won't break, it will just not work how you might expect. Again, if you don't like it, propose a solution.

File: src/javasources/KTool/src/org/kframework/backend/java/symbolic/PatternMatcher.java (hunk 17 above)
Comment (yilongli) [inline src/javasources/KTool/src/org/kframework/backend/java/symbolic/PatternMatcher.java:208]: I needed to provide an interface for matching two terms besides a term and a rule. So I took the fastest way to do this by exposing the constructor. You are right. I should have made another static method for this functionality. Thanks!

File: tests/regression/java-rewrite-engine/smt_model/program.test.out (hunk 14 above)
Comment (yilongli) [inline tests/regression/java-rewrite-engine/smt_model/program.test.out:2]: @andreistefanescu I need your help here. Is there any way to modify this test to address Dwight's last comment? Or could you somehow rewrite the test?

File: src/javasources/KTool/src/org/kframework/backend/java/symbolic/KILtoBackendJavaKILTransformer.java (hunk 16 above)
Comment (yilongli) [inline src/javasources/KTool/src/org/kframework/backend/java/symbolic/KILtoBackendJavaKILTransformer.java:20]: Okay, I can configure my Eclipse settings to do that.

File: tests/regression/java-rewrite-engine/smt_model/program.test.out (hunk 14 above)
Comment (dwightguth) [inline tests/regression/java-rewrite-engine/smt_model/program.test.out:2]: I'm sorry I was imprecise. I don't need this fixed necessarily. I need the
answer to the specific question I asked: is there a pattern you could write
which this test would match, but would nonetheless express the correctness
of this result?

File: src/javasources/KTool/src/org/kframework/backend/java/indexing/TokenIndex.java (hunk 15 above)
Comment (yilongli) [inline src/javasources/KTool/src/org/kframework/backend/java/indexing/TokenIndex.java]: This commit was made before your comment regarding quadratic number of code changes. For now, the right way to do this is to use `Token{"$h"}` instead of `"$h"` and write a kompiler pass to check it. The special case for KLabelIndex and TokenIndex won't be necessary then. The only special case is TopIndex and, thus, there will be no quadratic number of code changes.

**Comment by yilongli** [comment]
Right now, I don't have time to work on the code. I will let you know when it's ready for review again. Thanks.

File: tests/regression/java-rewrite-engine/smt_model/program.test.out (hunk 14 above)
Comment (yilongli) [inline tests/regression/java-rewrite-engine/smt_model/program.test.out:2]: No, I get your point. I was asking AndreiS if there is way to rewrite the test or specify a pattern in the ktest config file to get rid of this for good. This test is failing from time to time and I have to modify the output many times.

Comment (andreistefanescu) [inline tests/regression/java-rewrite-engine/smt_model/program.test.out:2]: This name is returned by Z3, so I have no control over it.

File: src/javasources/KTool/src/org/kframework/compile/transformers/AddInjections.java (hunk 13 above)
Comment (dwightguth) [inline src/javasources/KTool/src/org/kframework/compile/transformers/AddInjections.java:195]: @andreistefanescu , can you please clarify me in specific terms what the difference between an if_then_else_fi that we define and one the user would define would be?

File: tests/regression/java-rewrite-engine/smt_model/program.test.out (hunk 14 above)
Comment (dwightguth) [inline tests/regression/java-rewrite-engine/smt_model/program.test.out:2]: No, I get that we have an undefined variable name here. What I'm wondering is if it's possible to specify this test in such a way that we don't specify what the name of the variable needs to be?

Comment (yilongli) [inline tests/regression/java-rewrite-engine/smt_model/program.test.out:2]: Fixed by modifying the test to be insensitive w.r.t. the name of generated fresh variable.

**Comment by yilongli** [comment]
Ready for code review again.

File: src/javasources/KTool/src/org/kframework/compile/sharing/TokenSortCollector.java [hunk 18]
```diff
@@ -18,20 +25,82 @@
...
+     * Checks if a specified production is valid w.r.t. its sort: namely, a
+     * lexical production should have a token sort and a non-lexical production
+     * should have a non-token sort. Throws a K exception when one of the two
+     * restrictions is violated. Currently, this check is only enabled in the
+     * Java backend.
+     * 
+     * @param production
+     *            the specified production
+     */
+    private void checkIllegalProduction(Production production) {
```
Comment (dwightguth) [inline src/javasources/KTool/src/org/kframework/compile/sharing/TokenSortCollector.java:75]: Sorry just to make sure I understand this, we're not concerned about the case when a token is part of a sort that is a subsort of a sort containing KLabels? That doesn't cause anything to break if it happens?

**Comment by dwightguth** [comment]
This looks good, and Andrei answered my remaining questions. I am merging this.
